File: Java/Interpreter/BooleanExpressions/src/com/ericgtkb/BooleanExpression.java
package com.ericgtkb;

public interface BooleanExpression {
    boolean evaluate(Context context);

    default BooleanExpression and(BooleanExpression expression) {
        return new AndExpression(this, expression);
    }

    default BooleanExpression or(BooleanExpression expression) {
        return new OrExpression(this, expression);
    }

    default BooleanExpression not() {
        return new NotExpression(this);
    }
}

class VariableExpression implements BooleanExpression {
    private String name;

    public VariableExpression(String name) {
        this.name = name;
    }

    @Override
    public boolean evaluate(Context context) {
        return context.lookup(name);
    }

    public String getName() {
        return name;
    }

    @Override
    public String toString() {
        return name;
    }
}

class AndExpression implements BooleanExpression {
    private BooleanExpression expression1;
    private BooleanExpression expression2;

    public AndExpression(BooleanExpression expression1, BooleanExpression expression2) {
        this.expression1 = expression1;
        this.expression2 = expression2;
    }

    @Override
    public boolean evaluate(Context context) {
        return expression1.evaluate(context) && expression2.evaluate(context);
    }

    @Override
    public String toString() {
        return "(" + expression1 + " and " + expression2 + ")";
    }
}

class OrExpression implements BooleanExpression {
    private BooleanExpression expression1;
    private BooleanExpression expression2;

    public OrExpression(BooleanExpression expression1, BooleanExpression expression2) {
        this.expression1 = expression1;
        this.expression2 = expression2;
    }

    @Override
    public boolean evaluate(Context context) {
        return expression1.evaluate(context) || expression2.evaluate(context);
    }

    @Override
    public String toString() {
        return "(" + expression1 + " or " + expression2 + ")";
    }
}

class NotExpression implements BooleanExpression {
    private BooleanExpression expression;

    public NotExpression(BooleanExpression expression) {
        this.expression = expression;
    }

    @Override
    public boolean evaluate(Context context) {
        return !expression.evaluate(context);
    }

    @Override
    public String toString() {
        return "not " + expression;
    }
}

File: C++/Interpreter/BooleanExpressions/boolean_expression.cc
#include "boolean_expression.h"


std::ostream& operator<<(std::ostream& os, const BooleanExpression& expression) {
    expression._to_ostream(os);
    return os;
}

std::shared_ptr<BooleanExpression> operator&(const std::shared_ptr<BooleanExpression>& expression1,
                                             const std::shared_ptr<BooleanExpression>& expression2) {
    return std::make_shared<AndExpression>(expression1, expression2);
}

std::shared_ptr<BooleanExpression> operator|(const std::shared_ptr<BooleanExpression>& expression1,
                                             const std::shared_ptr<BooleanExpression>& expression2) {
    return std::make_shared<OrExpression>(expression1, expression2);
}

std::shared_ptr<BooleanExpression> operator~(const std::shared_ptr<BooleanExpression>& expression) {
    return std::make_shared<NotExpression>(expression);
}


VariableExpression::VariableExpression(const std::string& name) : _name(name) {}

const std::string& VariableExpression::get_name() const {
    return _name;
}

bool VariableExpression::evaluate(Context& context) {
    return context.lookup(_name);
}

void VariableExpression::_to_ostream(std::ostream& os) const {
    os << _name;
}


AndExpression::AndExpression(const std::shared_ptr<BooleanExpression>& expression1,
                             const std::shared_ptr<BooleanExpression>& expression2) : _expression1(expression1),
                                                                                      _expression2(expression2) {}

bool AndExpression::evaluate(Context& context) {
    return _expression1->evaluate(context) && _expression2->evaluate(context);
}

void AndExpression::_to_ostream(std::ostream& os) const {
    os << "(" << *_expression1 << " and " << *_expression2 << ")";
}


OrExpression::OrExpression(const std::shared_ptr<BooleanExpression>& expression1,
                             const std::shared_ptr<BooleanExpression>& expression2) : _expression1(expression1),
                                                                                      _expression2(expression2) {}

bool OrExpression::evaluate(Context& context) {
    return _expression1->evaluate(context) || _expression2->evaluate(context);
}

void OrExpression::_to_ostream(std::ostream& os) const {
    os << "(" << *_expression1 << " or " << *_expression2 << ")";
}


NotExpression::NotExpression(const std::shared_ptr<BooleanExpression>& expression) : _expression(expression) {}

bool NotExpression::evaluate(Context& context) {
    return !_expression->evaluate(context);
}

void NotExpression::_to_ostream(std::ostream& os) const {
    os << "not " << *_expression;
}
